File: apps/api/app/schemas/real/evidence_integrity.py
"""Evidence integrity schemas (real application layer, Phase 21).

Mirror the shapes the frontend consumes for evidence integrity while keeping
UUID identities and additive semantics. Reads never accept actor identity from
the client — the authenticated context supplies it.

Language is deliberately neutral: verification describes the integrity state of
a cryptographic digest, never the truthfulness of evidence content.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import Field

from app.schemas.common import SchemaBase


class CustodyEventRead(SchemaBase):
    sequence: int
    action: str
    event_timestamp: datetime | None = None
    evidence_checksum: str | None = None
    metadata_hash: str | None = None
    previous_event_hash: str | None = None
    current_event_hash: str | None = None
    actor: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class EvidenceIntegrityRead(SchemaBase):
    evidence_id: Any
    investigation_id: Any
    evidence_checksum: str | None = None
    custody_chain_hash: str | None = None
    associated_anchor_digest: str | None = None
    status: str  # COMPLETE | CHECKSUM_UNAVAILABLE | NOT_ANCHORED
    status_detail: str | None = None
    algorithm_version: str
    custody_events: list[CustodyEventRead] = Field(default_factory=list)
    generated_at: datetime


class AnchorViewRead(SchemaBase):
    anchor_id: Any
    evidence_id: Any
    investigation_id: Any
    custody_chain_hash: str
    anchor_digest: str
    network: str
    provider: str
    is_mock: bool
    status: str  # PENDING | ANCHORED | UNAVAILABLE
    transaction_id: str | None = None
    block_number: int | None = None
    contract_address: str | None = None
    anchored_at: datetime | None = None
    verified_at: datetime | None = None
    reason: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class EvidenceIntegrityProviderView(SchemaBase):
    provider: str
    network: str
    healthy: bool
    is_mock: bool
    detail: str


class EvidenceBlockchainRead(SchemaBase):
    """Full blockchain-anchor view for one evidence item (snake_case wire)."""

    evidence_id: Any
    investigation_id: Any
    integrity: EvidenceIntegrityRead
    anchor: AnchorViewRead | None = None
    provider: EvidenceIntegrityProviderView
    verification_state: str  # NOT_ANCHORED | VERIFIED | MISMATCH | PENDING | UNAVAILABLE
    last_verified_at: datetime | None = None
    message: str | None = None
    generated_at: datetime


class AnchorRequest(SchemaBase):
    """Optional client hints for the anchor action.

    No actor identity, no credentials, no evidence payload are accepted — the
    digest and custody hash are derived server-side only.
    """

    message: str | None = Field(default=None, max_length=200)


class AnchorResponse(SchemaBase):
    evidence_id: Any
    investigation_id: Any
    anchor: AnchorViewRead | None = None
    verification_state: str
    message: str | None = None
    anchored: bool = False
    already_anchored: bool = False


class VerifyResponse(SchemaBase):
    evidence_id: Any
    investigation_id: Any
    verification_state: str  # NOT_ANCHORED | VERIFIED | MISMATCH | PENDING | UNAVAILABLE
    current_custody_chain_hash: str | None = None
    anchored_custody_chain_hash: str | None = None
    on_chain_digest: str | None = None
    detail: str | None = None
    verified_at: datetime | None = None
    anchor: AnchorViewRead | None = None
    message: str | None = None

File: packages/types/src/evidence-integrity.ts
// ============================================================
// PHASE 21 — BLOCKCHAIN EVIDENCE INTEGRITY ANCHORING
// ============================================================
// Evidence integrity describes the cryptographic integrity state of an
// evidence custody chain and its optional blockchain anchor. Language is
// deliberately neutral: an anchored digest proves a checksum existed on-chain
// at a point in time; it proves nothing about whether the underlying evidence
// content is truthful, and it never implies guilt.
// ============================================================

export type EvidenceIntegrityVerificationState =
  | 'VERIFIED'
  | 'MISMATCH'
  | 'NOT_ANCHORED'
  | 'PENDING'
  | 'UNAVAILABLE';

export type EvidenceIntegrityStatus =
  | 'ANCHORED'
  | 'NOT_ANCHORED'
  | 'CHECKSUM_UNAVAILABLE';

/** One chained event of the derived evidence custody chain. */
export interface EvidenceCustodyEvent {
  sequence: number;
  action: string;
  eventTimestamp?: string | null;
  evidenceChecksum?: string | null;
  metadataHash?: string | null;
  previousEventHash?: string | null;
  currentEventHash?: string | null;
  actor?: string | null;
  metadata: Record<string, unknown>;
}

/** Derived custody-chain + checksum read for one evidence item. */
export interface EvidenceIntegritySummary {
  evidenceId: string;
  investigationId: string;
  evidenceChecksum?: string | null;
  custodyChainHash?: string | null;
  associatedAnchorDigest?: string | null;
  status: EvidenceIntegrityStatus;
  statusDetail?: string | null;
  algorithmVersion: string;
  custodyEvents: EvidenceCustodyEvent[];
  generatedAt: string;
}

/** Persisted blockchain anchor view (wire snake_case mapped). */
export interface EvidenceBlockchainAnchorView {
  anchorId: string;
  evidenceId: string;
  investigationId: string;
  custodyChainHash: string;
  anchorDigest: string;
  network: string;
  provider: string;
  isMock: boolean;
  status: 'PENDING' | 'ANCHORED' | 'UNAVAILABLE';
  transactionId?: string | null;
  blockNumber?: number | null;
  contractAddress?: string | null;
  anchoredAt?: string | null;
  verifiedAt?: string | null;
  reason?: string | null;
  metadata: Record<string, unknown>;
}

/** Blockchain provider health surfaced to the UI (never client secrets). */
export interface EvidenceIntegrityProviderView {
  provider: string;
  network: string;
  healthy: boolean;
  isMock: boolean;
  detail: string;
}

/** Full blockchain anchor + integrity view for one evidence item. */
export interface EvidenceBlockchainView {
  evidenceId: string;
  investigationId: string;
  integrity: EvidenceIntegritySummary;
  anchor?: EvidenceBlockchainAnchorView | null;
  provider: EvidenceIntegrityProviderView;
  verificationState: EvidenceIntegrityVerificationState;
  lastVerifiedAt?: string | null;
  message?: string | null;
  generatedAt: string;
}

/** Result of the anchor action (idempotent). */
export interface EvidenceAnchorResult {
  evidenceId: string;
  investigationId: string;
  anchor?: EvidenceBlockchainAnchorView | null;
  verificationState: EvidenceIntegrityVerificationState;
  message?: string | null;
  anchored: boolean;
  alreadyAnchored: boolean;
}

/** Result of the verify action. */
export interface EvidenceVerifyResult {
  evidenceId: string;
  investigationId: string;
  verificationState: EvidenceIntegrityVerificationState;
  currentCustodyChainHash?: string | null;
  anchoredCustodyChainHash?: string | null;
  onChainDigest?: string | null;
  detail?: string | null;
  verifiedAt?: string | null;
  anchor?: EvidenceBlockchainAnchorView | null;
  message?: string | null;
}

/** Compact integrity payload embedded in AI context (grounded, bounded). */
export interface EvidenceIntegrityContextPayload {
  verificationState: EvidenceIntegrityVerificationState;
  isMock: boolean;
  network?: string;
  anchorDigest?: string | null;
  custodyChainHash?: string;
  checksumPrefixed?: string | null;
}